db/db_impl/db_impl_open.cc

```diff
@@ -144,6 +145,23 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src,
     result.wal_dir = result.wal_dir.substr(0, result.wal_dir.size() - 1);
   }
 
+#ifdef OS_LINUX
+  // To fix https://github.com/facebook/rocksdb/issues/12038
+  if (!result.use_direct_reads && result.compaction_readahead_size > 0) {
+    size_t system_limit =
+        GetCompactionReadaheadSizeSystemLimit(result.db_paths);
+    if (system_limit > 0 && result.compaction_readahead_size > system_limit) {
+      result.compaction_readahead_size = system_limit;
```

Comment:
This could cause a regression on WSF. We should only do this for Posix.

Reply:
Good point - moved it to PosixFileSystem